File: backend/src/spotdl/api/v1/providers.py
"""Providers API endpoints."""

from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel

from spotdl.core.providers_config import (
    AUDIO_SOURCE_PROVIDERS,
    LYRICS_PROVIDERS,
    METADATA_PROVIDERS,
    ProviderInfo,
    get_default_preferences,
)

router = APIRouter(prefix="/providers")


class ProviderInfoResponse(BaseModel):
    """Provider information response."""

    id: str
    name: str
    icon: str
    default_enabled: bool


class ProvidersResponse(BaseModel):
    """Response model for all providers."""

    audio_sources: list[ProviderInfoResponse]
    metadata_sources: list[ProviderInfoResponse]
    lyrics_sources: list[ProviderInfoResponse]


class ProviderPreferenceResponse(BaseModel):
    """Provider preference response."""

    id: str
    enabled: bool


class DefaultPreferencesResponse(BaseModel):
    """Response model for default preferences."""

    audio: list[ProviderPreferenceResponse]
    metadata: list[ProviderPreferenceResponse]
    lyrics: list[ProviderPreferenceResponse]


@router.get("", response_model=ProvidersResponse)
async def list_providers() -> ProvidersResponse:
    """
    List all available providers grouped by category.

    Returns provider information including ID, display name, icon key,
    and whether the provider is enabled by default.
    """
    return ProvidersResponse(
        audio_sources=[ProviderInfoResponse(**p) for p in AUDIO_SOURCE_PROVIDERS],
        metadata_sources=[ProviderInfoResponse(**p) for p in METADATA_PROVIDERS],
        lyrics_sources=[ProviderInfoResponse(**p) for p in LYRICS_PROVIDERS],
    )


@router.get("/defaults", response_model=DefaultPreferencesResponse)
async def get_default_provider_preferences() -> DefaultPreferencesResponse:
    """
    Get default provider preferences for all categories.

    Returns the default order and enabled status for each provider
    in each category.
    """
    return DefaultPreferencesResponse(
        audio=[
            ProviderPreferenceResponse(**p) for p in get_default_preferences("audio")
        ],
        metadata=[
            ProviderPreferenceResponse(**p) for p in get_default_preferences("metadata")
        ],
        lyrics=[
            ProviderPreferenceResponse(**p) for p in get_default_preferences("lyrics")
        ],
    )

File: frontend/src/api/providers.ts
import { useQuery } from "@tanstack/react-query";
import { apiClient } from "./client";

export interface ProviderInfo {
  id: string;
  name: string;
  icon: string;
  default_enabled: boolean;
}

export interface ProvidersResponse {
  audio_sources: ProviderInfo[];
  metadata_sources: ProviderInfo[];
  lyrics_sources: ProviderInfo[];
}

export interface ProviderPreference {
  id: string;
  enabled: boolean;
}

export interface DefaultPreferencesResponse {
  audio: ProviderPreference[];
  metadata: ProviderPreference[];
  lyrics: ProviderPreference[];
}

export const providerKeys = {
  all: ["providers"] as const,
  list: () => [...providerKeys.all, "list"] as const,
  defaults: () => [...providerKeys.all, "defaults"] as const,
};

/**
 * Get all available providers grouped by category.
 */
export async function getProviders(): Promise<ProvidersResponse> {
  const response = await apiClient.get<ProvidersResponse>("/providers");
  return response.data;
}

/**
 * Get default provider preferences.
 */
export async function getDefaultProviderPreferences(): Promise<DefaultPreferencesResponse> {
  const response = await apiClient.get<DefaultPreferencesResponse>("/providers/defaults");
  return response.data;
}

/**
 * Hook to fetch all providers.
 */
export function useProviders() {
  return useQuery({
    queryKey: providerKeys.list(),
    queryFn: getProviders,
    staleTime: 1000 * 60 * 60, // 1 hour - providers rarely change
  });
}

/**
 * Hook to fetch default provider preferences.
 */
export function useDefaultProviderPreferences() {
  return useQuery({
    queryKey: providerKeys.defaults(),
    queryFn: getDefaultProviderPreferences,
    staleTime: 1000 * 60 * 60, // 1 hour
  });
}
